src/zenml/stack/stack_component.py

```diff
@@ -63,6 +67,156 @@ class StackComponent(BaseModel, ABC):
 
+    def __init__(self, **kwargs: Any) -> None:
+        """Ensures that secret references don't clash with pydantic validation.
+
+        StackComponents allow the specification of all their string attributes
+        using secret references of the form `{{secret_name.key}}`. This however
+        is only possible when the stack component does not perform any explicit
+        validation of this attribute using pydantic validators. If this were
+        the case, the validation would run on the secret reference and would
+        fail or in the worst case, modify the secret reference and lead to
+        unexpected behavior. This method ensures that no attributes that require
+        custom pydantic validation are set as secret references.
+
+        Args:
+            **kwargs: Arguments to initialize this stack component.
+
+        Raises:
+            ValueError: If an attribute that requires custom pydantic validation
+                is passed as a secret reference, or if the `name` attribute
+                was passed as a secret reference.
+        """
+        for key, value in kwargs.items():
+            try:
+                field = self.__class__.__fields__[key]
+            except KeyError:
+                # Value for a private attribute or non-existing field, this
+                # will fail during the upcoming pydantic validation
+                continue
+
+            if value is None:
+                continue
+
+            if not secret_utils.is_secret_reference(value):
+                if secret_utils.is_secret_field(field):
+                    logger.warning(
+                        "You specified a plain-text value for the sensitive "
+                        f"attribute `{key}` for a `{self.__class__.__name__}` "
+                        "stack component. This is currently only a warning, "
+                        "but future versions of ZenML will require you to pass "
+                        "in senstive information as secrets. Check out the "
+                        "documentation on how to configure your stack "
+                        "components with secrets here: "
+                        "https://docs.zenml.io/developer-guide/advanced-usage/secret-references"
+                    )
+                continue
+
+            if key == "name":
+                raise ValueError(
+                    "Passing the `name` attribute of a stack component as a "
+                    "secret reference is not allowed."
+                )
+
+            requires_validation = field.pre_validators or field.post_validators
+            if requires_validation:
+                raise ValueError(
+                    f"Passing the stack component attribute `{key}` as a "
+                    "secret reference is not allowed as additional validation "
+                    "is required for this attribute."
+                )
+
+        super().__init__(**kwargs)
+
+    def __custom_getattribute__(self, key: str) -> Any:
```

Comment:
This is perhaps my greatest fear related to this feature: that we somehow end up leaking actual secret values instead of references somewhere because we didn't cover a corner case.

Reply:
Yes this has been my biggest concern as well, as I thought having this distinction will be very complicated to implement. However any "serialization" related methods on pydantic models (e.g. `model.dict()`, `model.json()` or `dict(model`) internally don't directly access the attributes using `getattr(...)`, but instead iterate over `model.__fields__` or `model.__dict__` which contains the unresolved references. However you reminded me of a test that covers exactly this which I forgot to implement, will add that right away.